staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -77,4 +79,33 @@ type Response struct {
 	Audiences Audiences
 	// User is the UserInfo associated with the authentication context.
 	User user.Info
+	// AuthenticatorID identifies the authenticator that authenticated the request.
+	AuthenticatorID string
+}
+
+// Error is the error returned by an authenticator upon failures.
+type Error struct {
+	// Err is the error produced by an authenticator.
+	Err error
+	// AuthenticatorID identifies the authenticator that failed an authentication request.
+	AuthenticatorID string
+}
+
+// Error implements the errors.Error interface.
+func (a *Error) Error() string {
+	return fmt.Sprintf("authenticator: %q failed to process authentication request: err: %v", a.AuthenticatorID, a.Err)
+}
+
+// NewError constructs Error.
+// When the supplied error is nil, this method will also return nil.
+func NewError(authenticatorID string, err error) *Error {
+	if err == nil {
+		return nil
+	}
+	return &Error{AuthenticatorID: authenticatorID, Err: err}
+}
+
+// NewAggregatedError aggregates multiple errors into a single Error.
+func NewAggregatedError(authenticatorID string, errList []error) *Error {
```

Comment:
Can `NewError` take an error vararg instead?

Reply:
It could be.
The reason I used a slice is because this is how authenticator/token/union/union.go collects errors for aggregation.

Reply:
That authenticator can always pass the slice to vararg: `authenticator.NewError("union-token", errs...)`

Reply:
This function seems useless, can we just call utilerrors.NewAggregate before passing an error to NewError?